crates/language/src/language.rs

```diff
@@ -890,20 +892,28 @@ impl LanguageRegistry {
         })
     }
 
+    /// context_info is a (worktree ID, &AppContext),
+    /// which will be used to load user's custom file association settings.
     pub fn language_for_file(
         self: &Arc<Self>,
         path: impl AsRef<Path>,
+        context_info: Option<(usize, &AppContext)>,
         content: Option<&Rope>,
     ) -> UnwrapFuture<oneshot::Receiver<Result<Arc<Language>>>> {
         let path = path.as_ref();
         let filename = path.file_name().and_then(|name| name.to_str());
         let extension = path.extension_or_hidden_file_name();
         let path_suffixes = [extension, filename];
+        let associations = context_info.map(|(worktree_id, cx)| {
+            &AllLanguageSettings::get(Some((worktree_id, path)), cx).associations
+        });
         self.get_or_load_language(|config| {
-            let path_matches = config
-                .path_suffixes
-                .iter()
-                .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())));
+            // Give higher priority to user's custom settings
```

Comment:
So that means we have a way to totally mess with the default settings and rebind certain languages' suffixes to another ones?

Seems dangerous, if we want to go this way, we need more docs on the feature and the way it works then — so nobody has to guess; and we need tests with the fake LSP servers, to ensure that only one starts for such overrides.

Reply:
How about having a dedicate struct for file associations as I described [here](https://github.com/zed-industries/zed/pull/7344#issuecomment-1925790929)? It will give us a consistent  way to manage conflicts among default settings, global settings and local settings.

Reply:
Could work: as long as it's documented and is able to somehow emit errors on wrong settings.